# Predict if the server will be hacked

This is the project done for the Novartis Data Science Hiring Challenge conducted through HackerEarth.

### Problem Statement:

Nowadays, people often use their mode of payment through digital means. Due to this, the potential hackers who are hacking the account are also rising. The hackers can also take our important details just with the help of the phone number attached to the account.

### Dataset:

There is this data with some anonymized variables which can be used to predict whether the hack is going to happen or not.

The task is to build a predictive model which would help to predict whether the server will be hacked with much accuracy.

## Importing packages and reading the train data set

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [2]:
df_novartis = pd.read_csv(r"C:\Users\Srividhya\Desktop\Novartis datascience challenge\Dataset\Train.csv")

In [3]:
df_novartis.head()

INCIDENT_ID       DATE  X_1  X_2  X_3  X_4  X_5  X_6  X_7  X_8  X_9  X_10  \
0   CR_102659  04-JUL-04    0   36   34    2    1    5    6    1    6     1   
1   CR_189752  18-JUL-17    1   37   37    0    0   11   17    1    6     1   
2   CR_184637  15-MAR-17    0    3    2    3    5    1    0    2    3     1   
3   CR_139071  13-FEB-09    0   33   32    2    1    7    1    1    6     1   
4   CR_109335  13-APR-05    0   33   32    2    1    8    3    0    5     1   

   X_11  X_12  X_13  X_14  X_15  MULTIPLE_OFFENSE  
0   174   1.0    92    29    36                 0  
1   236   1.0   103   142    34                 1  
2   174   1.0   110    93    34                 1  
3   249   1.0    72    29    34                 1  
4   174   0.0   112    29    43                 1

## Exploratory Data Analysis - EDA of training data

In [4]:
df_novartis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23856 entries, 0 to 23855
Data columns (total 18 columns):
INCIDENT_ID         23856 non-null object
DATE                23856 non-null object
X_1                 23856 non-null int64
X_2                 23856 non-null int64
X_3                 23856 non-null int64
X_4                 23856 non-null int64
X_5                 23856 non-null int64
X_6                 23856 non-null int64
X_7                 23856 non-null int64
X_8                 23856 non-null int64
X_9                 23856 non-null int64
X_10                23856 non-null int64
X_11                23856 non-null int64
X_12                23674 non-null float64
X_13                23856 non-null int64
X_14                23856 non-null int64
X_15                23856 non-null int64
MULTIPLE_OFFENSE    23856 non-null int64
dtypes: float64(1), int64(15), object(2)
memory usage: 3.3+ MB


In [5]:
df_novartis.describe()

X_1           X_2           X_3           X_4           X_5  \
count  23856.000000  23856.000000  23856.000000  23856.000000  23856.000000   
mean       0.483778     24.791206     24.637450      4.276744      2.455609   
std        1.439738     15.240231     15.135093      2.944672      1.963095   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      7.000000      8.000000      2.000000      1.000000   
50%        0.000000     24.000000     24.000000      4.000000      3.000000   
75%        0.000000     36.000000     35.000000      6.000000      5.000000   
max        7.000000     52.000000     52.000000     10.000000      5.000000   

                X_6           X_7           X_8           X_9          X_10  \
count  23856.000000  23856.000000  23856.000000  23856.000000  23856.000000   
mean       6.154175      4.876509      0.972460      4.924128      1.244802   
std        4.471756      3.881931      1.453144      1.362625      1.119301   
min        1.000000      0.000000      0.000000      0.000000      1.000000   
25%        3.000000      2.000000      0.000000      5.000000      1.000000   
50%        5.000000      4.000000      1.000000      5.000000      1.000000   
75%        8.000000      7.000000      1.000000      6.000000      1.000000   
max       19.000000     18.000000     99.000000      6.000000     90.000000   

               X_11          X_12          X_13          X_14          X_15  \
count  23856.000000  23674.000000  23856.000000  23856.000000  23856.000000   
mean     206.954519      0.974064     85.237383     72.674296     33.464747   
std       93.033348      1.167725     27.597226     43.297320      8.386834   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%      174.000000      1.000000     72.000000     29.000000     34.000000   
50%      249.000000      1.000000     98.000000     62.000000     34.000000   
75%      249.000000      1.000000    103.000000    107.000000     34.000000   
max      332.000000     90.000000    116.000000    142.000000     50.000000   

       MULTIPLE_OFFENSE  
count      23856.000000  
mean           0.955231  
std            0.206800  
min            0.000000  
25%            1.000000  
50%            1.000000  
75%            1.000000  
max            1.000000

In [6]:
df_novartis.columns

Index(['INCIDENT_ID', 'DATE', 'X_1', 'X_2', 'X_3', 'X_4', 'X_5', 'X_6', 'X_7',
       'X_8', 'X_9', 'X_10', 'X_11', 'X_12', 'X_13', 'X_14', 'X_15',
       'MULTIPLE_OFFENSE'],
      dtype='object')

In [7]:
df_novartis.dtypes

INCIDENT_ID          object
DATE                 object
X_1                   int64
X_2                   int64
X_3                   int64
X_4                   int64
X_5                   int64
X_6                   int64
X_7                   int64
X_8                   int64
X_9                   int64
X_10                  int64
X_11                  int64
X_12                float64
X_13                  int64
X_14                  int64
X_15                  int64
MULTIPLE_OFFENSE      int64
dtype: object

###### Considering the above output of datatypes, it is clearly seen that the DATE column is not in the format of date but it is of the object datatype. So, it is necessary to convert the DATE column to its precise format.

In [8]:
df_novartis['DATE'] = pd.to_datetime(df_novartis['DATE'])

In [9]:
df_novartis.dtypes

INCIDENT_ID                 object
DATE                datetime64[ns]
X_1                          int64
X_2                          int64
X_3                          int64
X_4                          int64
X_5                          int64
X_6                          int64
X_7                          int64
X_8                          int64
X_9                          int64
X_10                         int64
X_11                         int64
X_12                       float64
X_13                         int64
X_14                         int64
X_15                         int64
MULTIPLE_OFFENSE             int64
dtype: object

###### Now, the DATE column looks accurate since it is converted into its precise datatype of datetime64[ns]

In [10]:
df_novartis.shape

(23856, 18)

In [11]:
df_novartis.isnull().sum()

INCIDENT_ID           0
DATE                  0
X_1                   0
X_2                   0
X_3                   0
X_4                   0
X_5                   0
X_6                   0
X_7                   0
X_8                   0
X_9                   0
X_10                  0
X_11                  0
X_12                182
X_13                  0
X_14                  0
X_15                  0
MULTIPLE_OFFENSE      0
dtype: int64

###### There are no null values in the train data except the X_12 column which has 182 missing data values. There are many approaches to deal with the missing values. Few are:
1. Dropping the entire column.
2. Filling the missing values with 0.
3. Filling the missing values with average value of the column or the most frequent value of the column.

Dropping the column is not a wise choice in this case because X_12 has huge range of values and dropping it would affect the model. Filling with 0 is also meaningless in this case.

###### Hence, I chose to fill the missing values with the most frequent values of the X_12 column.

In [12]:
df_novartis['X_12'].fillna(df_novartis['X_12'].mode()[0],inplace=True)

In [13]:
df_novartis.isnull().sum()

INCIDENT_ID         0
DATE                0
X_1                 0
X_2                 0
X_3                 0
X_4                 0
X_5                 0
X_6                 0
X_7                 0
X_8                 0
X_9                 0
X_10                0
X_11                0
X_12                0
X_13                0
X_14                0
X_15                0
MULTIPLE_OFFENSE    0
dtype: int64

###### Now, there are no null values in the train dataset.

In [14]:
df_novartis.duplicated().sum()

0

###### There are no duplicate values in the train dataset.

In [15]:
df_novartis.nunique()

INCIDENT_ID         23856
DATE                 9121
X_1                     8
X_2                    52
X_3                    52
X_4                    10
X_5                     5
X_6                    19
X_7                    19
X_8                    24
X_9                     7
X_10                   24
X_11                  133
X_12                   23
X_13                   60
X_14                   62
X_15                   28
MULTIPLE_OFFENSE        2
dtype: int64

In [16]:
df_novartis['MULTIPLE_OFFENSE'].value_counts()

1    22788
0     1068
Name: MULTIPLE_OFFENSE, dtype: int64

In [17]:
df_novartis['X_5'].value_counts()

5    7368
1    6818
3    4973
0    4695
2       2
Name: X_5, dtype: int64

In [18]:
df_novartis['X_9'].value_counts()

5    10559
6     9508
2     3040
3      452
1      175
0      118
4        4
Name: X_9, dtype: int64

In [19]:
df_novartis['X_1'].value_counts()

0    19036
1     3497
7      876
5      270
3      136
4       26
2       10
6        5
Name: X_1, dtype: int64

###### Now, the data is explored and preprocessed correctly for the model development, model fitting and predictions.

## Model Development

In [20]:
X = df_novartis.drop('MULTIPLE_OFFENSE',axis=1)

In [21]:
y = df_novartis.MULTIPLE_OFFENSE

In [22]:
X.head()

INCIDENT_ID       DATE  X_1  X_2  X_3  X_4  X_5  X_6  X_7  X_8  X_9  X_10  \
0   CR_102659 2004-07-04    0   36   34    2    1    5    6    1    6     1   
1   CR_189752 2017-07-18    1   37   37    0    0   11   17    1    6     1   
2   CR_184637 2017-03-15    0    3    2    3    5    1    0    2    3     1   
3   CR_139071 2009-02-13    0   33   32    2    1    7    1    1    6     1   
4   CR_109335 2005-04-13    0   33   32    2    1    8    3    0    5     1   

   X_11  X_12  X_13  X_14  X_15  
0   174   1.0    92    29    36  
1   236   1.0   103   142    34  
2   174   1.0   110    93    34  
3   249   1.0    72    29    34  
4   174   0.0   112    29    43

In [23]:
X.drop(['INCIDENT_ID','DATE'],axis = 1, inplace = True)

In [24]:
X.head()

X_1  X_2  X_3  X_4  X_5  X_6  X_7  X_8  X_9  X_10  X_11  X_12  X_13  X_14  \
0    0   36   34    2    1    5    6    1    6     1   174   1.0    92    29   
1    1   37   37    0    0   11   17    1    6     1   236   1.0   103   142   
2    0    3    2    3    5    1    0    2    3     1   174   1.0   110    93   
3    0   33   32    2    1    7    1    1    6     1   249   1.0    72    29   
4    0   33   32    2    1    8    3    0    5     1   174   0.0   112    29   

   X_15  
0    36  
1    34  
2    34  
3    34  
4    43

In [25]:
y.shape

(23856,)

In [26]:
X.shape

(23856, 15)

#### Importing the necessary packages and libraries

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

#### Model Fitting and making predictions on the train data

In this case, it is not wise to use simple models like LinearRegression, Logistic Regression, etc., since they may develop high error rates. So, the ensemble methods should be used. In my opinion, I used Random Forest method since our dataset contains anonymous data and random forest could classify correctly in that case.

In [55]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.20)
model1 = RandomForestRegressor(random_state=1)
model1.fit(X_train, y_train)
predictions1 = model1.predict(X_test)
scoree = mean_absolute_error(y_test, predictions1)
print('Mean Absolute Error score ',scoree)

Mean Absolute Error score  0.002682313495389774


###### The error rate of the trained model is 0.0026 which is lesser than the other simple models and hence this model can be used for predicting the test data.

## Exploratory Data Analysis - EDA on test data

In [56]:
test_novartis = pd.read_csv(r"C:\Users\Srividhya\Desktop\Novartis datascience challenge\Dataset\Test.csv")

In [57]:
test_novartis.head()

INCIDENT_ID       DATE  X_1  X_2  X_3  X_4  X_5  X_6  X_7  X_8  X_9  X_10  \
0   CR_195453  01-FEB-18    0   30   35    7    3    6    4    0    5     1   
1   CR_103520  05-MAR-04    0   44   44    1    3    7    1    4    6     1   
2   CR_196089  27-JAN-18    0   34   33    3    5    2    7    3    0     1   
3   CR_112195  18-AUG-06    7    3    2    3    5    9    8    0    5     1   
4   CR_149832  31-OCT-11    0    7    8    7    3    2    7    1    5     1   

   X_11  X_12  X_13  X_14  X_15  
0   174   NaN    72   119    23  
1   316   0.0    12    29    34  
2   316   1.0    72     0    34  
3   174   1.0   112    87    34  
4   174   0.0   112    93    43

In [58]:
test_novartis.isnull().sum()

INCIDENT_ID      0
DATE             0
X_1              0
X_2              0
X_3              0
X_4              0
X_5              0
X_6              0
X_7              0
X_8              0
X_9              0
X_10             0
X_11             0
X_12           127
X_13             0
X_14             0
X_15             0
dtype: int64

###### Similar to train data, the X_12 column of test data also contains null values. So, they can be approached similar to the approach did in train data. X_12 missing values are filled with the most frequent value in the same column.

In [59]:
test_novartis['X_12'].fillna(test_novartis['X_12'].mode()[0],inplace=True)

In [60]:
test_novartis.isnull().sum()

INCIDENT_ID    0
DATE           0
X_1            0
X_2            0
X_3            0
X_4            0
X_5            0
X_6            0
X_7            0
X_8            0
X_9            0
X_10           0
X_11           0
X_12           0
X_13           0
X_14           0
X_15           0
dtype: int64

In [61]:
test_novartis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15903 entries, 0 to 15902
Data columns (total 17 columns):
INCIDENT_ID    15903 non-null object
DATE           15903 non-null object
X_1            15903 non-null int64
X_2            15903 non-null int64
X_3            15903 non-null int64
X_4            15903 non-null int64
X_5            15903 non-null int64
X_6            15903 non-null int64
X_7            15903 non-null int64
X_8            15903 non-null int64
X_9            15903 non-null int64
X_10           15903 non-null int64
X_11           15903 non-null int64
X_12           15903 non-null float64
X_13           15903 non-null int64
X_14           15903 non-null int64
X_15           15903 non-null int64
dtypes: float64(1), int64(14), object(2)
memory usage: 2.1+ MB


In [62]:
test_novartis.shape

(15903, 17)

In [63]:
test_novartis.nunique()

INCIDENT_ID    15903
DATE            7939
X_1                8
X_2               52
X_3               52
X_4               10
X_5                5
X_6               19
X_7               19
X_8               22
X_9                7
X_10              18
X_11             106
X_12              17
X_13              54
X_14              51
X_15              29
dtype: int64

In [64]:
test_novartis['X_1'].value_counts()

0    12778
1     2264
7      550
5      188
3       92
4       22
2        7
6        2
Name: X_1, dtype: int64

In [65]:
test_novartis['X_4'].value_counts()

6     3581
2     3092
0     2253
7     1891
4     1342
3     1289
9      960
10     871
1      620
5        4
Name: X_4, dtype: int64

In [66]:
test_novartis['X_5'].value_counts()

5    4870
1    4434
3    3382
0    3213
2       4
Name: X_5, dtype: int64

In [67]:
test_novartis['X_9'].value_counts()

5    7051
6    6273
2    2051
3     310
1     135
0      82
4       1
Name: X_9, dtype: int64

In [68]:
test_novartis['DATE'] = pd.to_datetime(test_novartis['DATE'])

###### The datatype of the DATE column in the test data is also not correct. So, we changed it to its precise data type of datetime64[ns].

In [69]:
test_novartis.dtypes

INCIDENT_ID            object
DATE           datetime64[ns]
X_1                     int64
X_2                     int64
X_3                     int64
X_4                     int64
X_5                     int64
X_6                     int64
X_7                     int64
X_8                     int64
X_9                     int64
X_10                    int64
X_11                    int64
X_12                  float64
X_13                    int64
X_14                    int64
X_15                    int64
dtype: object

In [70]:
test_novartis.isnull().sum()

INCIDENT_ID    0
DATE           0
X_1            0
X_2            0
X_3            0
X_4            0
X_5            0
X_6            0
X_7            0
X_8            0
X_9            0
X_10           0
X_11           0
X_12           0
X_13           0
X_14           0
X_15           0
dtype: int64

The INCIDENT_ID and DATE columns have no use in predicting the value. So, they can be removed.

In [71]:
test1 = test_novartis.drop(['INCIDENT_ID', 'DATE'], axis = 1)

In [72]:
test1.head()

X_1  X_2  X_3  X_4  X_5  X_6  X_7  X_8  X_9  X_10  X_11  X_12  X_13  X_14  \
0    0   30   35    7    3    6    4    0    5     1   174   1.0    72   119   
1    0   44   44    1    3    7    1    4    6     1   316   0.0    12    29   
2    0   34   33    3    5    2    7    3    0     1   316   1.0    72     0   
3    7    3    2    3    5    9    8    0    5     1   174   1.0   112    87   
4    0    7    8    7    3    2    7    1    5     1   174   0.0   112    93   

   X_15  
0    23  
1    34  
2    34  
3    34  
4    43

## Predicting the target variable using the model developed

In [73]:
test_prediction = model1.predict(test1).round()

In [74]:
test_prediction[:20]

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 0., 1.])

In [75]:
output = pd.DataFrame({'INCIDENT_ID': test_novartis.INCIDENT_ID,
                        'MULTIPLE_OFFENSE': test_prediction})

In [76]:
output.to_csv(r"C:\Users\Srividhya\Desktop\Novartis datascience challenge\Dataset\submission4.csv",index=False)

## Conclusion:

Hence, the given data is clearly explored, cleaned and preprocessed to proceed to the model development. The RandomForestRegressor model also performed the best in predicting the training data. The test data is preprocessed similar to the train data and predictions are made using the model developed.